### PR TITLE
Add HTTPProxy FQDN Validation

### LINKED ISSUE #3166: HTTPProxy FQDN Validation
**Please describe the problem you have**

During the creation of new HTTPProxy resources, FQDNs are not validated resulting that invalid urls such as `my-domain-${somevariable}.domain.tld` are passed to envoy but the HTTPProxy is not created. Could the CRD definition be edited to include DNS1123 validation similar to [apimachinery](https://github.com/kubernetes/apimachinery/blob/2456ebdaba229616fab2161a615148884b46644b/pkg/util/validation/validation.go#L78)  to check that the FQDN does not container invalid characters?

### PR DESCRIPTION
**Pull request by abhide**
Fixes #3166

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -6,23 +6,23 @@ require (
...
-	github.com/stretchr/testify v1.5.1
+	github.com/stretchr/testify v1.6.1
 	google.golang.org/genproto v0.0.0-20200526211855-cb27e3aa2013
 	google.golang.org/grpc v1.27.0
 	google.golang.org/protobuf v1.25.0
 	gopkg.in/alecthomas/kingpin.v2 v2.2.6
 	gopkg.in/yaml.v2 v2.3.0
 	k8s.io/api v0.19.2
-	k8s.io/apimachinery v0.19.2
+	k8s.io/apimachinery v0.20.2
```
Comment (abhide) [inline go.mod:23]: Will separate ticket and send out a separate PR for go mod dependencies update

**Comment by abhide** [comment]
@skriss / @stevesloka Is this the correct place for adding the validation? Seems like there are lot of test failures with this validation.

File: internal/dag/status_test.go [hunk 2]
```diff
@@ -2369,4 +2369,36 @@ func TestDAGStatus(t *testing.T) {
...
+					Port: 8080,
+				}},
+			}},
+		},
+	}
+
+	run(t, "invalid FQDN", testcase{
+		objs: []interface{}{invalidFQDN, fixture.ServiceRootsHome},
+		want: map[types.NamespacedName]contour_api_v1.DetailedCondition{
+			{Name: invalidFQDN.Name, Namespace: invalidFQDN.Namespace}: fixture.NewValidCondition().
```
Comment (abhide) [inline internal/dag/status_test.go:2398]: intentional! want it to fail. Will fix as part of removing WIP

File: internal/dag/httpproxy_processor.go [hunk 3]
```diff
@@ -120,6 +121,13 @@ func (p *HTTPProxyProcessor) computeHTTPProxy(proxy *contour_api_v1.HTTPProxy) {
 			"Spec.VirtualHost.Fqdn must be specified")
 		return
 	}
+
+	// Check if FQDN is conformant according to RFC1123
+	if errs := validation.IsDNS1123Label(host); len(errs) != 0 {
```
Comment (youngnick) [inline internal/dag/httpproxy_processor.go:126]: The reason you're getting a lot of test failures here is that a DNS label is a single *component* of a hostname.

So, in `subdomain.foo.com`, `subdomain`, `foo`, and `com` are all labels. A period (`.`) is not permitted in a DNS label.

I'd suggest using `IsFullyQualifiedDomainName` instead, however, if it fails, you also need to check for a single DNS label (which we accept as valid).

So, something like:
```suggestion
	if errs := validation.IsFullyQualifiedDomainName(host); len(errs) != 0 {
	  if labelErr := validation.IsDNS1123Label; labelErr != nil {
	    // add the error
```
I think something like that should help clear up your errors, and then you just need to make sure that you're testing both some fully qualified names (like `subdomain.foo.com`, `foo.com`), and a single label (like `foo`).

**Review by youngnick — COMMENTED** [review]
I think you've run afoul of RFC-speak here, specifically about the difference between a *label* and a *hostname*.

File: internal/dag/httpproxy_processor.go (hunk 3 above)
Comment (abhide) [reply internal/dag/httpproxy_processor.go]: Thanks @youngnick . FIxed in  latest patch.

File: internal/dag/status_test.go (hunk 2 above)
Comment (abhide) [reply internal/dag/status_test.go:2470]: fixed

File: internal/dag/httpproxy_processor.go [hunk 4]
```diff
@@ -102,6 +103,28 @@ func (p *HTTPProxyProcessor) Run(dag *DAG, source *KubernetesCache) {
 	}
 }
 
+// Credits to https://github.com/kubernetes/apimachinery/blob/master/pkg/util/validation/validation.go#L95
```
Comment (abhide) [inline internal/dag/httpproxy_processor.go:106]: Is there a better way to use the original one from apimachinery?

File: internal/dag/httpproxy_processor.go [hunk 5]
```diff
@@ -102,6 +103,28 @@ func (p *HTTPProxyProcessor) Run(dag *DAG, source *KubernetesCache) {
...
+// Had to rewrite it here as we don't need FieldPath
+func (p *HTTPProxyProcessor) isValidFQDN(name string) []string {
```
Comment (youngnick) [inline internal/dag/httpproxy_processor.go:108]: This looks great! Unfortunately, because of Kube things, we actually need to allow a single DNS label as an option as well. I'm sorry not to have made that clear earlier. So, the check about having at least 2 segments can't be added, as it's an API change.

So, this function should check:
Is this a valid DNS subdomain (using the validation check)?
Are all the labels valid (one is a valid number of labels)?
If not, it's not valid.

For this new function, I'd also like to see unit tests that check each if statement adds the correct errors.

**Review by youngnick — CHANGES_REQUESTED** [review]
This is really close @abhide! Just a couple of things about the expected behavior, and a little more testing, and it will be done.
